src/orderly/cli/options.py

```diff
@@ -0,0 +1,130 @@
+import functools
+
+import click
+
+from outpack.search_options import SearchOptions
+
+
+class NumericalParamType(click.ParamType):
+    """
+    A type for numerical parameters.
+
+    It automatically deduces whether the parameter should be interpreted as an
+    int or a float.
+    """
+
+    name = "number"
+
+    def convert(self, value, param, ctx):
+        if isinstance(value, float):
+            return value
+
+        try:
+            return int(value)
+        except ValueError:
+            pass
+
+        try:
+            return float(value)
+        except ValueError:
+            self.fail(f"{value!r} is not a valid numerical value", param, ctx)
+
+
+# Almost all sub-commands accept a --root option, so it would make sense for it
+# to be accepted early in the command syntax. For instance, we want to accept
+# `orderly --root /path run foo`. Normally, with click, for this to be accepted
+# we would have to define the option on the top-level command.
+#
+# However there is at least one command (init) which does not take a --root
+# option, which would prevent us from doing this. Additionally, if the option
+# is defined on the top-level command, we cannot call the tool with the option
+# further down, ie. `orderly run --root /path foo`.
+#
+# To work around this we apply a bit of magic: we add a hidden option on the
```

Comment:
Would it be simpler to allow `--root` to be passed into the handler for init, then throw with something like `--root does not make sense for init`?

Reply:
That could work, but then it wouldn't support passing the root argument at the end, eg `orderly run foo --root /foo/bar`.